dev/ci/scripts/utils/rocotostat.py

```diff
@@ -3,22 +3,96 @@
 import sys
 import os
 import copy
-from time import sleep
+import logging
+from time import sleep, time
 
-from wxflow import which, Logger, CommandNotFoundError, ProcessError
+from wxflow import which, Logger, CommandNotFoundError, ProcessError, Executable
 from argparse import ArgumentParser, FileType
 
 from collections import Counter
 
-logger = Logger(level=os.environ.get("LOGGING_LEVEL", "DEBUG"), colored_log=False)
+logger = Logger(level=os.environ.get("LOGGING_LEVEL", "DEBUG"),
+                colored_log=False,
+                logfile_path=os.environ.get("ROCOTOSTAT_LOG_FILE"))
 
+# If logging to file, suppress console output unless explicitly requested
+if os.environ.get("ROCOTOSTAT_LOG_FILE") and not os.environ.get("ROCOTOSTAT_CONSOLE_LOG", "").lower() in ['true', '1', 'yes']:
+    # Remove console handlers to suppress stdout/stderr output when logging to file
+    for handler in logger._logger.handlers[:]:
+        if isinstance(handler, logging.StreamHandler) and handler.stream.name in ['<stdout>', '<stderr>']:
+            logger._logger.removeHandler(handler)
 
-def attempt_multiple_times(expression, max_attempts, sleep_duration=0, exception_class=Exception):
+
+def get_user_thread_count():
     """
-    Retries a function multiple times.
+    Get the current user thread count for monitoring against ulimit -u.
 
-    Try to execute the function expression up to max_attempts times ignoring any exceptions
-    of the type exception_class, It waits for sleep_duration seconds between attempts.
+    Returns
+    -------
+    dict
+        Dictionary containing thread count and limit information
+    """
+
+    try:
+        current_user = os.getenv('USER', 'unknown')
+
+        ps = which('ps')
+        try:
+            result = ps("-u", current_user, "-L", output=str)
+            user_threads = len(result.strip().split('\n')) - 1 if result.strip() else 0
+        except ProcessError:
+            user_threads = -1
+
+        bash = which('bash')
+        try:
+            result = bash("-c", "ulimit -u", output=str)
+            process_limit = int(result.strip())
+        except (ProcessError, ValueError):
+            process_limit = -1
+
+        return {
+            'user': current_user,
+            'thread_count': user_threads,
+            'process_limit': process_limit,
+            'utilization_pct': round((user_threads / process_limit * 100), 2) if process_limit > 0 else -1
+        }
+
+    except Exception as e:
+        logger.warning(f"Error getting user thread count: {e}")
+        return {
+            'user': 'unknown',
+            'thread_count': -1,
+            'process_limit': -1,
+            'utilization_pct': -1
+        }
+
+
+def log_thread_count(stage=""):
+    """
+    Log the current user thread count with optional stage identifier.
+
+    Parameters
+    ----------
+    stage : str
+        Optional stage identifier (e.g., "START", "END")
+    """
+    thread_info = get_user_thread_count()
```

Comment:
every call to `log_thread_count` is going to call `get_user_thread_count`.
The information received from `get_user_thread_count` is fixed and does not change.  Thread utilization will change, but which threads are being monitored here? `ps -u $USER -L` is all processes -- not just the ones that are of interest, and on that node.  If a job does get submitted to a different head-node, this information is not reliable and incomplete.

Reply:
Sure but we are always running this python script from within the pipeline via the bash wrapper script `check-ci.sh`  The pipeline runner is fixed on a specific head node, and is always ran "through" the moc bash script int a temp working space managed by the runner in the user account within the process that launched the runner executable. Therefor: 1) always ran within the parent process of the runner 2) this parent process is fixed on a specific head node.